README.md
# fracpane

A wrapper for [tweakpane](https://github.com/cocopon/tweakpane) that I've hacked together for personal use.


CHANGELOG.md
# @fractalhq/fracpane

## 0.2.1

### Patch Changes

- [fix] types / id parsing / docs ([`646b692`](https://github.com/FractalHQ/fracpane/commit/646b692fda83e4bf9e43f7ed33958076965c0be6))

## 0.2.0

### Minor Changes

- [feat] export svelte component ([`8052e6b`](https://github.com/FractalHQ/fracpane/commit/8052e6be1bfc42abe71a25c1edf5e6ae45d980c9))

## 0.1.5

### Patch Changes

- [fix] changeset bug ([`bd5fcb5`](https://github.com/FractalHQ/fracpane/commit/bd5fcb500e87c940e757e07a2e339891fff97d25))

## 0.1.4

### Patch Changes

- [fix] deps ([`4213e82`](https://github.com/FractalHQ/fracpane/commit/4213e82afbbd490380671ce6d0aadf121ba46146))

## 0.1.3

### Patch Changes

- [fix] move peers to deps ([`9d7eb0c`](https://github.com/FractalHQ/fracpane/commit/9d7eb0cd9b2d20495718e1ade3e5c48859073e42))

## 0.1.2

### Patch Changes

- [fix] peer deps ([`86834a2`](https://github.com/FractalHQ/fracpane/commit/86834a2fa18b04fa0d590351101171db7de8e010))

## 0.1.1

### Patch Changes

- [fix] dist ([`12febe5`](https://github.com/FractalHQ/fracpane/commit/12febe5aae5f2d8d49c17c8851421b6549caab2a))

## 0.1.0

### Minor Changes

- [feat] smart snapping / window resize reactivity ([`5ffc76a`](https://github.com/FractalHQ/fracpane/commit/5ffc76aba1d54932935e946b2cf04c70c0c94976))

## 0.0.3

### Patch Changes

- [release] remove scope ([`8093e3d`](https://github.com/FractalHQ/fracpane/commit/8093e3ddff02d6af2ae90f453e5e4ee66923e805))

## 0.0.2

### Patch Changes

- [release] initial ([`679a8b7`](https://github.com/FractalHQ/fracpane/commit/679a8b77a09d86026def8b32b6e34f1c3aeb78f9))


svelte.config.js
import mdsvexConfig from './mdsvex.config.js'
import adapter from '@sveltejs/adapter-auto'
import preprocess from 'svelte-preprocess'
import { mdsvex } from 'mdsvex'
import 'dotenv/config'

import { readFileSync } from 'fs'
import { fileURLToPath } from 'url'

const file = fileURLToPath(new URL('package.json', import.meta.url))
const json = readFileSync(file, 'utf8')
const pkg = JSON.parse(json)

/** @type {import('@sveltejs/kit').Config} */
const config = {
	extensions: ['.svelte', '.md'],
	preprocess: [
		preprocess({
			postcss: true,
		}),
		mdsvex(mdsvexConfig),
	],

	kit: {
		adapter: adapter(),
		prerender: {
			enabled: false, // SSR only for security
		},
		alias: {
			$package: 'src/package',
		},
	},

	package: {
		source: 'src/package', // Library directory
		dir: 'dist', // Output directory
		exports: (file) => file === 'index.ts', // Only export index.ts
		// external: Object.keys(pkg.peerDependencies), // External dependencies
	},

	vitePlugin: {
		experimental: {
			// https://github.com/sveltejs/vite-plugin-svelte/blob/main/docs/config.md#inspector
			inspector: {
				toggleButtonPos: 'bottom-left',
				toggleKeyCombo: 'meta-shift',
				showToggleButton: 'active',
				holdMode: true,
			},
		},
	},
}

export default config


src/package/utils/resizable.ts
import { persistentAtom } from '@nanostores/persistent'
import { debounce } from './debounce'
import { Subject } from 'rxjs'

export interface ResizeOptions {
	side?: 'left' | 'right' | 'top' | 'bottom'
	/**
	 * The starting width of the element.
	 * @default 420
	 */
	initialWidth?: string
	/**
	 * The size of the resize handle in pixels.
	 * @default '5px'
	 */
	gutterSize?: number | string
	/**
	 * The minimum size of the element in px.
	 * @default 50
	 */
	minSize?: number
	/**
	 * The maximum size of the element in px.
	 * @default 1000
	 */
	maxSize?: number
	/**
	 * Use a visible or invisible gutter.
	 * @default true
	 */
	visible?: boolean
	/**
	 * Gutter css color (if visible = `true`)
	 * @default 'transparent'
	 */
	color?: string
	/**
	 * Persist width in local storage.
	 * @default true
	 */
	persistent?: boolean
	/**
	 * The key to use for local storage.  Important if you have multiple resizable panes.
	 * @default 'fracpane_size'
	 */
	id?: string
}

const px = (size: number | string) => {
	if (typeof size === 'number') return `${size}px`
	else return size
}

export const resize = (node: HTMLElement, options: ResizeOptions) => {
	const {
		side = 'left',
		initialWidth = '420px',
		gutterSize = '3px',
		minSize = 50,
		maxSize = 1000,
		persistent = true,
		visible = true,
		color = 'transparent',
		id = 'fracpane',
	} = options

	const onResize = new Subject<{ delta: number; direction: string }>()

	const sizeStore = persistentAtom<string>(id + '_size', initialWidth, {
		encode: JSON.stringify,
		decode: JSON.parse,
	})

	const grabber = document.createElement('div')
	grabber.classList.add('resize-grabber')
	node.appendChild(grabber)
	node.style.cssText += `padding-${side}: ${px(gutterSize)} !important;`

	if (persistent) {
		const persistentSize = sizeStore.get()
		node.style.width = persistentSize
	}

	const direction = side === ('top' || 'bottom') ? 'vertical' : 'horizontal'
	const width = direction === 'horizontal' ? px(gutterSize) : px(node.scrollWidth)
	const height = direction === 'horizontal' ? px(node.clientHeight + node.scrollTop) : px(gutterSize)

	direction === 'horizontal' ? (grabber.style.top = '0') : (grabber.style.left = '0')
	const cursor = (grabber.style.cursor = direction === 'horizontal' ? 'ew-resize' : 'ns-resize')

	grabber.classList.add('grabber_' + side)
	grabber.style.cssText += `
		${side}: 0;
		width: ${width};
		height: ${height};
		background: ${visible ? color : 'transparent'};
		padding: ${px(gutterSize)};
		border-${direction === 'horizontal' ? 'top' : 'left'}-${side}-radius: 3px;
		border-${direction === 'horizontal' ? 'bottom' : 'right'}-${side}-radius: 3px;
		position: absolute;
		display: flex;
		flex-grow: 1;
		z-index: -1;
		cursor: ${cursor};
	`

	const onGrab = (e: MouseEvent) => {
		grabber.style.cursor = 'grabbing'
		const styleEl = document.createElement('style')
		styleEl.innerHTML = `
			* { cursor: grabbing !important; }
		`
		document.head.appendChild(styleEl)
		e.preventDefault()
		e.stopPropagation()
		document.addEventListener('mousemove', onMove)
		document.addEventListener('mouseup', (e) => onUp(e, styleEl), { once: true })
	}
	grabber.addEventListener('mousedown', onGrab)

	node.addEventListener('mouseover', updateHeight)

	const onMove = (e: MouseEvent) => {
		const startWidth = node.offsetWidth
		const startHeight = node.offsetHeight

		const deltaX = e.movementX
		const deltaY = e.movementY

		const minSizePx = minSize
		const maxSizePx = maxSize

		let newWidth = 0
		if (side === 'left') newWidth = startWidth - deltaX
		else if (side === 'right') newWidth = startWidth + deltaX

		const newHeight = startHeight + deltaY

		if (direction === 'horizontal') {
			if (newWidth < minSizePx) return
			if (maxSizePx && newWidth > maxSizePx) return
			node.style.width = `${newWidth}px`
			onResize.next({ delta: deltaX, direction: 'horizontal' })
		} else {
			if (newHeight < minSizePx) return
			if (maxSizePx && newHeight > maxSizePx) return
			node.style.height = `${newHeight}px`
			onResize.next({ delta: deltaY, direction: 'vertical' })
		}

		debounce(() => {
			sizeStore.set(px(node.offsetWidth))
		}, 50)
	}

	const onUp = (e: Event, styleEl?: HTMLStyleElement) => {
		document.removeEventListener('mousemove', onMove)
		// Remove the cursor style element.
		if (styleEl) document.head.removeChild(styleEl)
		grabber.style.cursor = cursor
	}

	document.addEventListener('mouseup', onUp)

	function updateHeight() {
		const node_rect = node.getBoundingClientRect()
		direction === 'horizontal'
			? (grabber.style.height = px(Math.min(node.scrollHeight, node_rect.height)))
			: (grabber.style.width = px(Math.min(node.scrollWidth, node_rect.width)))
	}

	setTimeout(updateHeight, 20)

	// Add a resize observer to update the height of the grabber.
	const resizeObserver = new ResizeObserver(updateHeight)
	resizeObserver.observe(node)

	return {
		onResize,
		destroy: () => {
			resizeObserver.disconnect()
			document.removeEventListener('mousemove', onMove)
			document.removeEventListener('mouseup', onUp)
			node.removeEventListener('mouseover', updateHeight)
			grabber.removeEventListener('mousedown', onGrab)
			grabber.remove()
		},
	}
}


src/package/fracpane.ts
import type { WritableAtom } from 'nanostores'
import type { Pane } from 'tweakpane'

import { Draggable, type DragOptions } from '@neodrag/vanilla'
import { resize, type ResizeOptions } from './utils/resizable'
import { persistentAtom } from '@nanostores/persistent'
import { log } from 'fractils'

export type Position = {
	x: number
	y: number
}

// todo this class would be much better as a factory function
export class Fracpane {
	private static instances = 0
	private static stylesAdded = false
	private id: string

	container?: HTMLElement
	element?: HTMLElement
	css?: string

	pane!: Pane
	title = 'controls'

	persistent = true
	position?: Position
	positionStore?: WritableAtom<Position>

	dragInstance?: Draggable
	dragging = false
	dragTimer?: NodeJS.Timeout | null = null

	minSize = 300
	width = 420

	dragOptions: DragOptions = {
		handle: '.tp-rotv_b',
		bounds: 'body',
		onDrag: () => {
			this.dragging = true
		},

		// Prevent dragging from collapsing gui
		onDragEnd: (e) => {
			if (this.dragging) {
				if (this.pane) this.pane.expanded = !this.pane.expanded

				this.dragTimer = setTimeout(() => {
					this.dragging = false

					// Update the position in localstorage if persistent
					this.positionStore?.set({
						x: e.offsetX,
						y: e.offsetY,
					})

					// Fixes bug where dragInstance.options.y is not updated
					this.dragInstance?.updateOptions({ position: { x: e.offsetX, y: e.offsetY } })
				}, 300)
			}
		},
	}

	resizerLeft?: ReturnType<typeof resize>
	resizerRight?: ReturnType<typeof resize>

	resizeOptions: ResizeOptions = {
		minSize: this.minSize,
		initialWidth: this.width + 'px',
	}

	windowWidth = 0
	windowHeight = 0

	disposing = false

	constructor(public options: Partial<Fracpane>) {
		Object.assign(this, options)

		if (!Fracpane.stylesAdded) {
			this.addStyles()
			Fracpane.stylesAdded = true
		}

		Fracpane.instances++

		if (!options.title && Fracpane.instances > 1) this.title += `_${Fracpane.instances}`

		this.id = `fracpane_${Fracpane.instances}_${this.title.trim().replace(/\s/g, '_')}`

		this.position ??= { x: 10, y: 10 }

		if (this.persistent) {
			this.positionStore = persistentAtom<Position>(`${this.id}_position`, this.position, {
				encode: JSON.stringify,
				decode: JSON.parse,
			})
			this.position = this.positionStore.get()
		}
	}

	public async init() {
		if (typeof window === 'undefined') return console.warn('skipping fracpane server-side')

		this.windowWidth = window.innerWidth
		this.windowHeight = window.innerHeight

		const needsMounting = !this.container

		this.container ??= document.createElement('div')
		this.container.classList.add(this.id + '_container')
		this.container.classList.add('fracpane')

		this.element ??= document.createElement('div')
		this.element.classList.add(this.id + '_element')

		const dragOptions = this.dragOptions
		// Update the position from localstorage if it exists
		if (this.persistent) this.dragOptions.position = this.position
		this.dragInstance = new Draggable(this.container, dragOptions)

		this.resizerLeft = resize(this.container, {
			id: this.id,
			side: 'left',
			gutterSize: '5px',
			initialWidth: this.width + 'px',
			...this.resizeOptions,
		})

		// Add a right resize handle
		this.resizerRight = resize(this.container, {
			id: this.id,
			side: 'right',
			gutterSize: '5px',
			initialWidth: this.width + 'px',
			...this.resizeOptions,
		})

		let updatePositionTimer: NodeJS.Timeout
		this.resizerLeft.onResize.subscribe(({ delta, direction }) => {
			if (this.dragInstance) {
				const currentPosition = {
					x: this.dragInstance.options.position?.x ?? 0,
					y: this.dragInstance.options.position?.y ?? 0,
				}
				if (direction === 'horizontal') {
					const newPosition = { x: currentPosition.x + delta, y: currentPosition.y }
					this.dragInstance.updateOptions({ position: newPosition })
				}
				if (direction === 'vertical') {
					const newPosition = { x: currentPosition.x, y: currentPosition.y + delta }
					this.dragInstance.updateOptions({ position: newPosition })
				}
				clearTimeout(updatePositionTimer)
				updatePositionTimer = setTimeout(() => {
					this.positionStore?.set({
						x: this.dragInstance?.options.position?.x ?? 0,
						y: this.dragInstance?.options.position?.y ?? 0,
					})
				}, 100)
			}
		})

		const Pane = (await import('tweakpane')).Pane

		this.pane = new Pane({
			title: this.title,
			container: this.container,
			expanded: true,
		})

		// Register plugins just in case
		const plugins = await import('@tweakpane/plugin-essentials')
		this.pane.registerPlugin(plugins)

		this.container.appendChild(this.pane.element)

		if (needsMounting) document.body.appendChild(this.container)

		// MutationRecord.removedNodes
		const onDestroy: MutationCallback = (mutationList, observer) => {
			mutationList.forEach((mutation) => {
				if (mutation.removedNodes.length) {
					mutation.removedNodes.forEach((node) => {
						if (node === this.container) {
							this.maybeDispose()
							observer.disconnect()
						}
					})
				}
			})
		}

		const observer = new MutationObserver(onDestroy)
		observer.observe(document.body, { childList: true })

		this.dragInstance?.updateOptions({ position: this.position })

		window.addEventListener('beforeunload', this.maybeDispose)
		window.addEventListener('resize', this.onWindowResize)

		this.onWindowResize()

		return this
	}

	dispose() {
		this.disposing = true
		log('disposing', this.id, 'sandybrown')
		window.removeEventListener('beforeunload', this.maybeDispose)
		window.removeEventListener('resize', this.onWindowResize)
		this.resizerLeft?.destroy()
		this.resizerRight?.destroy()
		this.dragInstance?.destroy()
		this.pane?.dispose()
		this.element?.remove()
		this.container?.remove()
		Fracpane.instances--
	}

	maybeDispose = () => {
		if (this.disposing) return
		this.dispose()
	}

	async addStyles() {
		const styles = import.meta.glob('./styles.css', { as: 'raw' })
		const styleTag = document.createElement('style')
		styleTag.innerHTML = await styles['./styles.css']()
		document.head.appendChild(styleTag)
	}

	onWindowResize = () => {
		if (this.dragInstance) {
			const lastWindowWidth = this.windowWidth
			const lastWindowHeight = this.windowHeight
			const currentWindowWidth = window.innerWidth
			const currentWindowHeight = window.innerHeight
			const currentWidth = this.container?.clientWidth ?? this.minSize
			const currentHeight = this.container?.clientHeight ?? this.minSize
			const { x, y } = this.dragInstance.options.position ?? { x: 0, y: 0 }

			// Find the ratio between the last window size and the current window size
			const widthRatio = currentWindowWidth / lastWindowWidth
			const heightRatio = currentWindowHeight / lastWindowHeight

			// Multiply the current position by the ratio to get the new position
			let newX = x * widthRatio
			let newY = y * heightRatio

			// Keep the container locked to the edge if it's placed close to the edge
			// On the left
			const lockThreshold = 50
			if (x < lockThreshold) newX = x
			if (y < lockThreshold) newY = y
			// On the right
			if (x > lastWindowWidth - currentWidth - lockThreshold) {
				// Stay locked to the right, but maintain the current distance from the right edge
				const distanceFromRightEdge = lastWindowWidth - x - currentWidth
				newX = currentWindowWidth - distanceFromRightEdge - currentWidth
			}
			if (y > lastWindowHeight - currentHeight - lockThreshold) {
				newY = currentWindowHeight - currentHeight
			}

			// If the new position is outside the window, move it to the edge
			if (newX > currentWindowWidth - currentWidth) {
				newX = currentWindowWidth - currentWidth
			}
			if (newY > currentWindowHeight - currentHeight) {
				// If the new position is outside the window, move the bottom of the container to the bottom of the window
				newY = currentWindowHeight - currentHeight
			}

			// Update the position
			this.dragInstance.updateOptions({ position: { x: newX, y: newY } })

			// Update the window size
			this.windowWidth = currentWindowWidth
			this.windowHeight = currentWindowHeight
		}
	}
}

export const fracpane = async (options?: Partial<Fracpane>): Promise<Fracpane> => {
	return (await new Fracpane(options ?? {}).init()) as Fracpane
}
